### PR TITLE
#54648: create copy of gpt-oss for quasar testing

### LINKED ISSUE #54648: Create copy of gpt-oss for quasar porting

### PR DESCRIPTION
**Pull request by bbradelTT**
### PR Category

Feature, Test Only

### Summary

Closes #54648 

Copies gpt-oss to a separate directory for quasar testing. Adds in a layer env var to allow running on a device with less SRAM.

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:bbradel-54648_gpt_oss_qsr)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:bbradel-54648_gpt_oss_qsr)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:bbradel-54648_gpt_oss_qsr)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:bbradel-54648_gpt_oss_qsr)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:bbradel-54648_gpt_oss_qsr)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:bbradel-54648_gpt_oss_qsr)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:bbradel-54648_gpt_oss_qsr)

### CODE REVIEW

File: models/experimental/ops/quasar/gpt_oss/tests/accuracy/test_model.py [hunk 1]
```diff
@@ -0,0 +1,386 @@
...
+
+
+@parametrize_mesh_with_fabric()
+def test_full_model_accuracy(mesh_device, device_params, reset_seeds, state_dict):
+    """Test full model with accuracy testing using new abstractions"""
+
+    # Cache file for reference tokens
+    cache_dir = "models/experimental/ops/quasar/gpt_oss/tests/accuracy/"
+    os.makedirs(cache_dir, exist_ok=True)
+    cache_file = os.path.join(cache_dir, "reference_tokens.pkl")
```
[the remark this answers is not in the document]
Comment (bbradelTT) [reply models/experimental/ops/quasar/gpt_oss/tests/accuracy/test_model.py:148]: pkl file deleted.

File: models/experimental/ops/quasar/gpt_oss/tt/common.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    gpt_oss_model_args = ModelArgs(
+        mesh_device,
+        max_batch_size=max_batch_size,
+        optimizations=optimizations,
+        max_seq_len=max_seq_len,
+    )
+    # Override num_layers if provided (useful for quick testing with fewer layers)
+    if num_layers is not None:
+        gpt_oss_model_args.hf_config.num_hidden_layers = num_layers
+        gpt_oss_model_args.n_layers = num_layers
```
[the remark this answers is not in the document]
Comment (bbradelTT) [reply models/experimental/ops/quasar/gpt_oss/tt/common.py:57]: Addressed

File: models/experimental/ops/quasar/gpt_oss/tests/run_logprobs_tests.sh [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+#   --demo-compare     Run both demos, save output to files, print perf diff
+#
+# Environment variables:
+#   HF_MODEL         Path to gpt-oss-120b weights (default: /localdev/gpt-oss-120b)
+#   TT_CACHE_PATH    Path to TT cache (default: /localdev/gpt-oss-120b)
+
+set -e
+
+# Navigate to tt-metal repo root (script lives in models/experimental/ops/quasar/gpt_oss/tests/)
+cd "$(dirname "$0")/../../../.."
```
[the remark this answers is not in the document]
Comment (bbradelTT) [reply models/experimental/ops/quasar/gpt_oss/tests/run_logprobs_tests.sh]: Addressed.

File: models/experimental/ops/quasar/gpt_oss/README.md [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+# GPT-OSS-120B (higher quality, requires more memory)
+export HF_MODEL="/mnt/MLPerf/tt_dnn-models/openai/gpt-oss-120b"
+```
+
+## Testing
+
+```bash
+# Run all tests
+pytest models/experimental/ops/quasar/gpt_oss/tests/unit/ -v
```
[the remark this answers is not in the document]
Comment (bbradelTT) [reply models/experimental/ops/quasar/gpt_oss/README.md:67]: Tests need to be added separately.

File: models/experimental/ops/quasar/gpt_oss/tt/experts_throughput/prefill.py [hunk 5]
```diff
@@ -0,0 +1,508 @@
...
+    if w_5d.layout != ttnn.ROW_MAJOR_LAYOUT:
+        w_5d_prev = w_5d
+        w_5d = ttnn.to_layout(w_5d, ttnn.ROW_MAJOR_LAYOUT, memory_config=ttnn.DRAM_MEMORY_CONFIG)
+        w_5d_is_new = w_5d is not w_5d_prev
+    try:
+        summed = ttnn.experimental.deepseek_prefill.post_combine_reduce(
+            combined, w_5d, output_memory_config=ttnn.DRAM_MEMORY_CONFIG
+        )
+    except Exception as _e:
+        logger.warning(f"FUSED FALLBACK: {_e}")
```
[the remark this answers is not in the document]
Comment (bbradelTT) [reply models/experimental/ops/quasar/gpt_oss/tt/experts_throughput/prefill.py:474]: Same as upstream. Keeping.

File: models/experimental/ops/quasar/gpt_oss/tests/fused_op_unit_tests/test_gpt_oss_all_fused_ops_device_perf.py [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+from models.perf.benchmarking_utils import BenchmarkData, BenchmarkProfiler
+
+
+def pytest_addoption(parser):
+    """Add custom command-line options for the test."""
+    parser.addoption(
+        "--fused-ops",
+        action="store",
+        default=None,
+        help="Comma-separated list of fused ops to test (default: all)",
```
[the remark this answers is not in the document]
Comment (bbradelTT) [reply models/experimental/ops/quasar/gpt_oss/tests/fused_op_unit_tests/test_gpt_oss_all_fused_ops_device_perf.py:54]: Same as upstream. Keeping.

File: models/experimental/ops/quasar/gpt_oss/tests/accuracy/test_model.py [hunk 7]
```diff
@@ -0,0 +1,386 @@
...
+    logger.info(f"Token 0 - Ref model: '{ref_token_text}' (id: {reference_tokens[0].item()})")
+
+    # For subsequent tokens, use decode (not prefill) like the demo
+    # Initialize like the demo
+    all_outputs = [input_tokens_prefill_pt[0][:prefill_len]]  # Start with prefill tokens
+    all_outputs[0].append(predicted_token)  # Add first generated token
+
+    # Initialize decode state like the demo
+    current_pos = torch.tensor([prefill_len])  # Start position after prefill
+    out_tok = torch.tensor([[predicted_token]])  # Current token to decode
```
[the remark this answers is not in the document]
Comment (bbradelTT) [reply models/experimental/ops/quasar/gpt_oss/tests/accuracy/test_model.py:102]: Same as upstream. Keeping.

File: models/experimental/ops/quasar/gpt_oss/tests/accuracy/test_model.py [hunk 8]
```diff
@@ -0,0 +1,386 @@
...
+
+        (
+            model_args,
+            model,
+            page_table,
+            tt_kv_cache,
+            tokenizer,
+            processor,
+            paged_attention_config,
+        ) = prepare_gpt_oss_generator_args(
```
[the remark this answers is not in the document]
Comment (bbradelTT) [reply models/experimental/ops/quasar/gpt_oss/tests/accuracy/test_model.py:213]: Same as upstream. Keeping.

File: models/experimental/ops/quasar/gpt_oss/tests/accuracy/test_model.py [hunk 9]
```diff
@@ -0,0 +1,386 @@
...
+        # Create reference model for comparison (use full model like demo)
+        from transformers.models.gpt_oss.modeling_gpt_oss import GptOssForCausalLM
+
+        # Load the same weights that the TTNN model uses
+        logger.info("Loading reference model weights...")
+        reference_weights = setup["model_args"].load_state_dict(
+            weights_path=setup["model_args"].model_path,
+            dummy_weights=setup["model_args"].dummy_weights,
+            convert_to_meta_format=True,
+        )
```
[the remark this answers is not in the document]
Comment (bbradelTT) [reply models/experimental/ops/quasar/gpt_oss/tests/accuracy/test_model.py:184]: Same as upstream. Keeping.

File: models/experimental/ops/quasar/gpt_oss/tt/experts/prefill.py [hunk 10]
```diff
@@ -0,0 +1,342 @@
...
+            is_input_a_sparse=True,
+            program_config=program_config.get_prefill_down_config(
+                down_input_split.shape[2], weights.down_proj.shape[-1], k=down_input_split.shape[-1]
+            ),
+            dtype=activation_dtype,
+        )
+        down_input_split.deallocate(True)
+
+        # Apply bias and routing weights
+        split_seq_len = seq_len if seq_len < split_size else split_size
```
[the remark this answers is not in the document]
Comment (bbradelTT) [reply models/experimental/ops/quasar/gpt_oss/tt/experts/prefill.py:199]: Same as upstream. Keeping.

File: models/experimental/ops/quasar/gpt_oss/config.py [hunk 11]
```diff
@@ -0,0 +1,232 @@
...
+        if config.tp > tp_dim_size:
+            raise ValueError(f"{mode.value}: TP({config.tp}) > mesh_{self.tp_axis}_size({tp_dim_size})")
+
+        # EP>1 makes the expert dimension sharded across ep_axis, and ttnn.moe_routing_remap
+        # requires expert_parallel_size to equal that axis extent exactly. Catch it here rather
+        # than as a device-side TT_FATAL mid-forward. EP=1 is unconstrained: prefill runs EP=1
+        # on multi-row meshes and never reaches the remap.
+        ep_dim_size = self.mesh_shape[self.ep_axis]
+        if config.ep > 1 and config.ep != ep_dim_size:
+            raise ValueError(f"{mode.value}: EP({config.ep}) != mesh_{self.ep_axis}_size({ep_dim_size})")
```
[the remark this answers is not in the document]
Comment (bbradelTT) [reply models/experimental/ops/quasar/gpt_oss/config.py:96]: Same as upstream. Keeping.

**Review by edwinleeTT — APPROVED** [review]
